### PR TITLE
awscloud: try to terminate previous secure instance (HMS-4341)

### PR DESCRIPTION
**Pull request by croissanne**
In case the previous executor SI belonging to the worker did not get shut down properly, attempt to do it again when starting a new one, otherwise replacing the SG or LT will not work.

### CODE REVIEW

File: internal/cloud/awscloud/secure-instance.go [hunk 1]
```diff
@@ -78,6 +78,11 @@ func (a *AWS) RunSecureInstance(iamProfile, keyName, cloudWatchGroup, hostname s
 		}
 	}()
 
+	previousSI, err := a.terminatePreviousSI(identity.InstanceID)
+	if err != nil {
+		logrus.Warningf("Unable to terminate previous secure instance with instance ID (%s): %v, this job is now likely to fail", previousSI, err)
```
Comment (schuellerf) [inline internal/cloud/awscloud/secure-instance.go:83]: logging `previousSI` in this case is not really useful in all cases? 
or maybe ... (see below)

File: internal/cloud/awscloud/secure-instance.go [hunk 2]
```diff
@@ -194,6 +199,42 @@ func (a *AWS) TerminateSecureInstance(si *SecureInstance) error {
...
+	})
+	if err != nil {
+		return "", err
+	}
+	if len(descrInstancesOutput.Reservations) == 0 || len(descrInstancesOutput.Reservations[0].Instances) == 0 {
+		return "", nil
+	}
+
+	if *descrInstancesOutput.Reservations[0].Instances[0].State.Name == ec2.InstanceStateNameTerminated {
+		return "", nil
```
Comment (schuellerf) [inline internal/cloud/awscloud/secure-instance.go:219]: … maybe change this for better error messages
```suggestion
		return "instance not yet terminated", nil
```

File: internal/cloud/awscloud/secure-instance.go [hunk 3]
```diff
@@ -194,6 +199,42 @@ func (a *AWS) TerminateSecureInstance(si *SecureInstance) error {
...
+				Name:   aws.String("tag:parent"),
+				Values: []*string{aws.String(hostInstanceID)},
+			},
+		},
+	})
+	if err != nil {
+		return "", err
+	}
+	if len(descrInstancesOutput.Reservations) == 0 || len(descrInstancesOutput.Reservations[0].Instances) == 0 {
+		return "", nil
```
Comment (schuellerf) [inline internal/cloud/awscloud/secure-instance.go:215]: … maybe change this for better error messages
```suggestion
		return "instance has no reservations", nil
```

File: internal/cloud/awscloud/secure-instance.go [hunk 4]
```diff
@@ -194,6 +199,42 @@ func (a *AWS) TerminateSecureInstance(si *SecureInstance) error {
...
+	descrInstancesOutput, err := a.ec2.DescribeInstances(&ec2.DescribeInstancesInput{
+		Filters: []*ec2.Filter{
+			&ec2.Filter{
+				Name:   aws.String("tag:parent"),
+				Values: []*string{aws.String(hostInstanceID)},
+			},
+		},
+	})
+	if err != nil {
+		return "", err
```
Comment (schuellerf) [inline internal/cloud/awscloud/secure-instance.go:212]: … maybe change this for better error messages
```suggestion
		return "no instance found", err
```

**Review by schuellerf — CHANGES_REQUESTED** [review]

File: internal/cloud/awscloud/secure-instance.go [hunk 5]
```diff
@@ -78,6 +78,11 @@ func (a *AWS) RunSecureInstance(iamProfile, keyName, cloudWatchGroup, hostname s
 		}
 	}()
 
+	previousSI, err := a.terminatePreviousSI(identity.InstanceID)
+	if err != nil {
+		logrus.Warningf("Unable to terminate previous secure instance with instance ID (%s): %v, this job is now likely to fail", previousSI, err)
+	}
```
Comment (schuellerf) [inline internal/cloud/awscloud/secure-instance.go:84]: would it be also interesting as a metric to print the terminated `previousSI` in the `else` case too?
```suggestion
	} else {
	  logrus.Infof("Previous secure instance with instance ID (%s) terminated", previousSI)
	}
```

File: internal/cloud/awscloud/secure-instance.go (hunk 1 above)
Comment (croissanne) [reply internal/cloud/awscloud/secure-instance.go:83]: ah right, I think we should replace it with the host/worker id maybe

File: internal/cloud/awscloud/secure-instance.go (hunk 5 above)
Comment (croissanne) [reply internal/cloud/awscloud/secure-instance.go:84]: sure

File: internal/cloud/awscloud/secure-instance.go (hunk 4 above)
Comment (croissanne) [reply internal/cloud/awscloud/secure-instance.go:212]: Well in this case the describe instances call failed, so it's not that there were no instances found.

File: internal/cloud/awscloud/secure-instance.go (hunk 2 above)
Comment (croissanne) [reply internal/cloud/awscloud/secure-instance.go:219]: n this case the instance is already terminated, which is good

File: internal/cloud/awscloud/secure-instance.go (hunk 3 above)
Comment (croissanne) [reply internal/cloud/awscloud/secure-instance.go:215]: in this case there are no active executor instances, which is fine. I don't think it should be logged.

**Review by schuellerf — APPROVED** [review]
LGTM

**Pull request merged**